modules/core/src/main/java/org/locationtech/jts/triangulate/VoronoiDiagramBuilder.java

```diff
@@ -171,7 +172,7 @@ private static Geometry clipGeometryCollection(Geometry geom, Envelope clipEnv)
 			if (clipEnv.contains(g.getEnvelopeInternal()))
 					result = g;
 			else if (clipEnv.intersects(g.getEnvelopeInternal())) {
-				result = clipPoly.intersection(g);
+				result = OverlayOp.intersection(clipPoly, g);
```

Comment:
Is this needed to avoid a dependency cycle?

Seems unfortunate if so, in that it's odd that library classes can't use it's actual API.

Reply:
Yes it's to avoid dependency cycle. Java is lenient on circular dependencies (by making it mandatory that Java compilers to do multi-pass compilation) but that doesn't make it a good idea. That said, I'm also not sure what a good design would be.

Reply:
> That said, I'm also not sure what a good design would be.

You mean a better design to avoid a dependency cycle?

This seems like the simplest way to avoid a cycle.  But it introduces a different dependency, on an implementation class which might change (e.g. if OverlayNG is used as the new overlay code).  And if a dynamic method registry is used this would have be changed as well.

Reply:
I mean I do not have a clear picture of a better design to decouple geometry and operations that would work well even with transpilation in mind. I don't know how to do "runtime dynamic method registry" without reflection in Java. In this case in particular reflection not good for portability/transpilation, but I also have a general problem with it as it bypasses the security of Java as a static typed language and makes it harder to do safe refactoring.

Reply:
I'm thinking that geometry operations would be reified (turned into classes with a generic interface for running operations).  The current `Op` class convention was a non-functional step in that direction back before I really had a solid idea about what to do. 

Alternatively, each general operation (overlay, relate, etc) will provide an interface exposing the operations it provides.  A registry will hold instances of these interfaces.

Would one of these work for transpilation?  Although not sure I want to make that big a change right now.

Reply:
I can see where you are going and it might be a sound design, but I can't imagine it at a good enough level to say which direction would be nicest. As for meaningful transpilation of  interfaces it's undiscovered country (for me).